aidial_adapter_bedrock/dial_api/storage.py

```diff
@@ -41,30 +63,85 @@ def to_form_data(
         return data
 
     async def upload(
-        self, filename: str, content_type: str, content: bytes
+        self,
+        directories: List[str],
+        filename: str,
+        content_type: str,
+        content: bytes,
     ) -> FileMetadata:
         async with aiohttp.ClientSession() as session:
-            data = FileStorage.to_form_data(filename, content_type, content)
-            async with session.post(
-                self.base_url,
+            data = FileStorage._to_form_data(filename, content_type, content)
+            async with session.put(
+                url="/".join([self.upload_url, *directories, filename]),
                 data=data,
-                headers=self.auth_headers(),
+                headers=self.auth.headers,
             ) as response:
                 response.raise_for_status()
                 meta = await response.json()
                 log.debug(
-                    f"Uploaded file: path={self.base_url}, file={filename}, metadata={meta}"
+                    f"Uploaded file: path={self.upload_url}, file={filename}, metadata={meta}"
                 )
                 return meta
 
+    async def upload_file_as_base64(
+        self, data: str, content_type: str
+    ) -> FileMetadata:
+        path_segments = _compute_path_as_hash(data)
+        directories, filename = path_segments[:-1], path_segments[-1]
+
+        ext = _get_extension(content_type)
+        filename = f"{filename}.{ext}" if ext is not None else filename
+
+        content: bytes = base64.b64decode(data)
+        return await self.upload(directories, filename, content_type, content)
+
 
-def _hash_digest(string: str) -> str:
-    return hashlib.sha256(string.encode()).hexdigest()
+def _compute_path_as_hash(file_content: str) -> List[str]:
```

Comment:
I think we don't need this step for case when we save results to the user folder. The user will not generate so many files that this could affect performance, but on the other hand, he may be afraid by such a directory structure that is difficult to navigate.

Reply:
Removed